firestore-send-email/functions/src/index.ts

```diff
@@ -278,76 +272,146 @@ async function deliver(
 
   // Wrapping in transaction to allow for automatic retries (#48)
   return admin.firestore().runTransaction((transaction) => {
+    // We could check state here is still PROCESSING, but we don't
+    // since the email sending will have been attempted regardless of what the
+    // delivery state was at that point, so we just update the state to reflect
+    // the result of the last attempt so as to not potentially cause duplicate sends.
     transaction.update(ref, update);
     return Promise.resolve();
   });
 }
 
-async function processWrite(change) {
-  if (!change.after.exists) {
-    return null;
-  }
+async function processWrite(
+  change: functions.Change<FirebaseFirestore.DocumentSnapshot<QueuePayload>>
+): Promise<void> {
+  const ref = change.after.ref;
+
+  // A quick check to avoid doing unnecessary transaction work.
+  // If the record state is SUCCESS or ERROR we don't need to do anything
+  // transactionally here since these are the 'final' delivery states.
+  // Note: we still check these again inside the transaction in case the state has
+  // changed while the transaction was inflight.
+  if (change.after.exists) {
+    const payloadAfter = change.after.data();
```

Comment:
Let's say F1 and F2 are instances of the same event handler that are created at almost the same time (handler was triggered more than once for a single event). F1 and F2's change.after.data() will both show delivery.state=PENDING, even if F1 already updated delivery.state=PROCESSING.

Perhaps one way to address that is for both F1 and F2 to ignore change.after.data() and read the latest document directly from Firestore. There's still a race condition if F2 reads the latest document before F1 has committed the transaction, but the current implementation AFAICT guarantees that a duplicate function call will deliver twice.

Reply:
We only use change.after here in this one check and only to check for an already finalized status (status that should never change) - it's only here to avoid unnecessary transaction work, since a status of "SUCCESS" or "ERROR" should never change (and if by chance it should change externally by the user or something then it'd trigger another function anyway which would skip past this check)

Reply:
Gotcha. Just making sure I understand how the duplicate logic works with an example, call it F1 and F2 which are duplicate instances of the same Firestore trigger.

F1 will put the document in PENDING state. When F2 calls deliver(), it will see that the document is in PENDING state, this it exits early. Am I understanding this correctly?